js/src/common/extend.ts

```diff
@@ -19,23 +26,27 @@
  *   // something that needs to be run on creation and update
  * });
  *
- * @param {object} object The object that owns the method
- * @param {string|string[]} methods The name or names of the method(s) to extend
- * @param {function} callback A callback which mutates the method's output
+ * @param object The object that owns the method
+ * @param methods The name or names of the method(s) to extend
+ * @param callback A callback which mutates the method's output
  */
-export function extend(object, methods, callback) {
+export function extend<T extends object, K extends KeyOfType<T, Function>>(
+  object: T,
+  methods: K | K[],
+  callback: (this: T, val: ReturnType<T[K]>, ...args: Parameters<T[K]>) => void
```

Comment:
Typescript recognizes that `this` is `this`? Cool!

On a side note, I really wish PHP had callback typing...

Reply:
Yep! You can type this